Bugfix/compile msm all matlab code

tbbrown opened the pull request:
Created new script to compile all MATLAB code: `global/matlab/Compile_All_MATLAB_code.sh`

* This script uses the existing versions of `Compile_MATLAB_code.sh` in the `ICAFIX/scripts` and `MSMAll/scripts` directories to compile the ICAFIX-related MATLAB code and the MSMAll-related MATLAB code respectively.
* A new script (`RestingStateStats/Compile_MATLAB_code.sh`) was created to compile the RestingStateStats-related MATLAB code. As with the previous `Compile_MATLAB_code.sh` scripts, this script can be used independently to just compile the RestingStateStats-related MATLAB code, but it is also invoked by the `Compile_All_MATLAB_code.sh` script.

Additional related changes include:

* Updated existing `Compile_MATLAB_code.sh` scripts to use the `log_Check_Env_Var` function
* Updated the `LICENSE.md` file to add 2019 to the copyright years and reference the CCF
* Removed project file (`RestingStateStats/RestingStateStats.prj`) which was only used for compiling the RestingStateStats MATLAB code from within the MATLAB GUI. This is no longer necessary since this compiling will be done via the `RestingStateStats/Compile_MATLAB_code.sh`. 
* Changed `RestingStateStats/RestingStateStats.sh` to use the `run_RestingStateStats.sh` produced by the new way of compiling the RestingStateStats MATLAB code (the `run_*` code is no longer found in a `distrib` subdirectory).

Important Notes:

* All MATLAB code compiled in this branch were compiled with MATLAB release R2016b.
* The resulting RestingStateStats compiled binaries **have not** been test run! This has been left to be done when RestingStateStats is overhauled to work with the results of the new MR-FIX.
* When RestingStateStats is overhauled to work with the results of the new MR-FIX, we should consider moving the `RestingStateStats.m` code along with the `RestingStateStats/Compile_MATLAB_code.sh` down into a subdirectory named `RestingStateStats/scripts` as is done for both the ICAFIX-related and MSMAll-related MATLAB code. This will require changes to both the  `global/matlab/Compile_All_MATLAB_code.sh` file and the `RestingStateStats.sh` file.

coalsont commented on RestingStateStats/Compiled_RestingStateStats/readme.txt:
```diff
@@ -2,27 +2,27 @@ MATLAB Compiler
+  have installed version 9.1 (R2016b).   
 
-. If the MCR is not installed, do the following:
+. If the MATLAB Runtime is not installed, do the following:
```
This should probably say "If you have this version of MATLAB installed, but the MATLAB Runtime is not installed".

tbbrown replied:
This `readme.txt` file is generated by the MATLAB compiler. So we don't have control over it to change it's contents.

coalsont commented on RestingStateStats/RestingStateStats.sh:
```diff
@@ -957,13 +957,13 @@ main()
 			# TBD: Use environment variable instead of fixed path
 			if [ "${CLUSTER}" = "1.0" ]; then
-				matlab_compiler_runtime="/export/matlab/R2013a/MCR"
+				matlab_compiler_runtime="/export/matlab/R2016b/MCR"
```
SetUpHCPPipeline.sh has MATLAB_COMPILER_RUNTIME in it, which should probably be used as-is here.  You can modify the wustl-specific SetUp... to have the CLUSTER variable logic in it, and thus remove CLUSTER from all the pipelines scripts.

tbbrown replied:
Thanks. Updated to use MATLAB_COMPILER_RUNTIME.

The CLUSTER variable is not likely to be needed anymore. It was introduced when the CHPC was in the change over from the "old" cluster to the "new" cluster and things were installed in different locations depending upon the cluster. The "old" cluster is dead, and changing the value of the MATLAB_COMPILER_RUNTIME variable in the SetUpHCPPipeline.sh file should be sufficiently flexible.

The CLUSTER variable is also likely to be confusing to external users.

tbbrown replied:
I think that eliminates the last remaining use of the CLUSTER variable in the pipeline scripts.

tbbrown commented:
To run the compilation of the MATLAB code, you have to have the 
MATLAB_HOME environment variable set to contain the home directory of 
your MATLAB installation, the MATLAB installation must have the MATLAB 
compiler option/package installed, and you have to have a license for 
the MATLAB compiler. For example, my MATLAB _HOME is set to 
/usr/local/MATLAB/R2016b. To compile with the latest MATLAB, I would 
need to install the latest version along with the compiler option. I 
already have the license set up to use the campus-wide pooled license 
that allows some maximum number of uses of the compiler license at once. 
(I don't know off the top of my head what that maximum number is.)

As for the RestingStateStats, I would have to look through my email to 
see if I received that development version. If I don't find it, I'll ask 
you to email it again. Is that development version ready to be put in 
the repository and replace the existing version?

On 5/13/19 20:19, glasserm wrote:
>
> *@glasserm* commented on this pull request.
>
> What are the external requirements to run this code? What if one 
> wanted to run with the latest matlab?
>
> As for RestingStateStats, I believe I have e-mailed around the 
> development version of this that works with MR+FIX and can be used for 
> applying temporal ICA cleanup.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Washington-University/HCPpipelines/pull/112?email_source=notifications&email_token=ABNXRRZGRSUFSJEX3DIW2KTPVIHRJA5CNFSM4HMSAVB2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYQCM7Q#pullrequestreview-236988030>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABNXRR7C7HQI764PVVBIF4DPVIHRJANCNFSM4HMSAVBQ>.
>

glasserm commented:
At this point it probably should not be a dropin replacement, but rather one of the files should be renamed...

mharms commented:
Isn't the dev version of RSS available here
https://github.com/Washington-University/HCPpipelines/tree/rss-mrfix
or is that not truly the latest dev version?

mharms commented:
Also, should we perhaps add some comments to `Compile_All_MATLAB_code.sh` along the lines of your response above regarding what is necessary to compile the matlab code? i.e., It may be obvious to us that you need the compiler option/package and license, but might be helpful to some if we stated that explicitly?

glasserm commented:
I haven't edited RSS dev since then.  I'm fine with the other suggestions I think.

glasserm commented:
Is this ready for final review and merge?

tbbrown commented:
I've changed the default MATLAB mode for the RestingStateStats to 1 (interpreted MATLAB) and moved the `RestingStateStats.m` MATLAB source code and corresponding `Compile_MATLAB_code.sh` script for RestingStateStats into a `scripts` sub-directory.

@mharms I'm not clear what other things are needed to "bring the manner in which interpreted matlab/octave are invoked into alignment with the ICAFIX and MSMAll related scripts". As it stands now, I believe:

`hcp_fix_multi_run`

1. Uses the `FSL_FIX_MATLAB_MODE` environment variable to determine whether to use compiled MATLAB, interpreted MATLAB, or interpreted OCTAVE
2. Has no default MODE. If `FSL_FIX_MATLAB_MODE` is not set, the script reports and error and aborts. 

`hcp_fix`

1. Shows the `FSL_FIX_MATLAB_MODE` environment variable, but doesn't seem to actually use it.
2. Seems that `fix` itself invoked within `hcp_fix` must make use of the `FSL_FIX_MATLAB_MODE` environment variable.

`PostFix.sh`, `ReApplyFixPipeline.sh`, `ReApplyFixMultiRunPipeline.sh`, `MSMAllPipeline.sh`, and `RestingStateStats.sh` all

1. Use the `--matlab-run-mode=` command line option to determine whether to use compiled MATLAB, interpreted MATLAB, or interpreted OCTAVE.
2. Default to mode 1, use interpreted MATLAB.

Are you thinking about using the `"${interpreter[@]}" <<< "$matlabCode"` approach that is used in `MSMAll.sh` and `SingleSubjectConcat.sh` in the `RestingStateStats.sh`?

tbbrown commented:
@glasserm 

I don't think it is quite ready for final review and merge.

mharms commented on RestingStateStats/RestingStateStats.sh:
```diff
@@ -993,7 +992,7 @@ main()
 			touch ${matlab_script_file_name}
-			echo "addpath ${HCPPIPEDIR}/RestingStateStats " >> ${matlab_script_file_name}
+			echo "addpath ${HCPPIPEDIR}/scripts/RestingStateStats " >> ${matlab_script_file_name}
 			echo "addpath /home/HCPpipeline/pipeline_tools/gifti" >> ${matlab_script_file_name}
```
We should fix this line to not call an irrelevant hard-coded path, but rather `addpath ${HCPPIPEDIR}/global/matlab`.  Note that because I previously added a symlink there of `@gifti@ -> gifti-1.6/@gifti`, adding `${HCPPIPEDIR}/global/matlab` to the matlab path has the effect of automatically adding both the cifti{open,save,savereset}.m functions to the matlab path and their needed gifti support as well (the latter being the intended purpose of L996).

mharms replied:
Also, L995 is wrong.  It should be `addpath ${HCPPIPEDIR}/RestingStateStats/scripts`, or use something like the `addpath('${this_script_dir}/scripts')` construction that we've added to the other matlab-related scripts, to make it more robust against directory name changes.

tbbrown replied:
* I've adapted `RestingStateStats.sh` to use similar techniques for invoking MATLAB or Octave to the other scripts.
* I also added comments to `Compile_All_MATLAB_code.sh` noting that users really should not need to do this.
* I fixed the use of `addpath` statements in `RestingStateStats.sh` to no longer contain a hard coded path to the gifti code
* I fixed the mistaken use of `${HCPPIPEDIR}/scripts/RestingStateStats` to `${HCPPIPEDIR}/RestingStateStats/scripts` in the interpreted code section

* I'm unclear on whether there has been a decision that we should replace this RestingStateStats code with code from the `rss-mrfix` branch or whether the code in the `rss-mrfix` branch should be merged into this branch with different file names so that it will end up added to the `master` branch when this pull request is merged, but still have separate file names (So both this code and the code from the `rss-mrfix` branch will co-exist.)

glasserm replied:
Probably need to stay separate until the code is actually merged.

mharms commented on global/matlab/Compile_All_MATLAB_code.sh:
```diff
@@ -22,6 +22,83 @@
+
+  This script can be used to (re)-compile all MATLAB code for functions called by
+  any pipelines that are part of the HCP Pipelines.
+
```
Maybe make even more explicit, early in the usage/comments, that use of this script is only necessary if someone wants/needs to RE-compile all the matlab-related scripts.  i.e., that the distribution _already provides_ compiled matlab binaries for use with one specific MCR version, namely R2016b/v91.

mharms commented:
@tbbrown Yes, I was referring to the `"${interpreter[@]}" <<< "$matlabCode"` approach, and collapsing of modes 1 and 2 into a single `1 | 2)` case branch -- both now used consistently throughout most of the scripts that actually invoke matlab -- i.e., `PostFix.sh, ReApplyFixPipeline.sh, ReApplyFixMultiRunPipeline.sh, MSMAll.sh, and SingleSubjectConcat.sh`.  

`hcp_fix` and `hcp_fix_multi_run` are exceptions to this, due to their legacy connections with the version of `hcp_fix` originally provided by the FIX distribution itself.

mharms commented:
Once we have what we think is "near-final", I'm happy to test that the compiled matlab mode actually runs to completion.

mharms commented:
So, with `--bc-mode=NONE`, RSS runs to completion if using Octave, and `--wm=NONE` and `--csf=NONE` (which, incidentally, are the default values for those two flags if not specified).  It does not run to completion under Octave if those are not NONE, in which case the problem isn't with WM and CSF code per se, but rather that the installation of Octave that I was using (4.2.1) choked on the grayplot generation:

```
error: called from
    graphics_toolkit at line 88 column 9
    figure at line 86 column 7
    RestingStateStats>plotgray at line 726 column 1
    RestingStateStats at line 352 column 5
```
Getting the figure (png) generation working under Octave could be a major undertaking, and I don't plan on pursuing that further at this time.

When attempting to run in compiled matlab mode, it abended with the following:

```
Undefined function or variable 'save_avw'.
Error in RestingStateStats (line 143)
MATLAB:UndefinedFunction
```

@tbbrown needs to recompile RestingStateStats.m after adding `-a ${HCPPIPEDIR}/global/fsl/etc/matlab` to the compile script, at which point I'll try running it again.

I just pushed a handful of changes into this branch.

mharms commented:
@tbbrown Just to make sure you didn't miss my comment that I've edited the branch, please update your local branch before you edit the compile script (which I didn't do) and re-compile.  Thx.

mharms commented:
@glasserm  Can you make sure to check 65e6e1a ?  I think it is preferable for RSS to no longer automatically move the _vn file that it generates (thus overwriting one that may already be in place in `ResultsFolder`), but the vn stuff is hard to keep track of across the various scripts.

Regarding the vn file generated by RSS:  Isn't it the case that the vn file generated by FIX (both single-run and multi-run) is computed _on top of_ the bias field correction?  Except if one runs RSS with `bc-mode=CORRECT`, it appears that the "real_bias" field is _backed out_. https://github.com/Washington-University/HCPpipelines/blob/86f4c4881ff3cad8a41f201e5d6a5077969024cc/RestingStateStats/scripts/RestingStateStats.m#L517

Is that an accurate statement of what is occurring?

glasserm commented:
The the idea will be to not compute the _vn files with RSS any more.  Perhaps they should be renamed _vn_RSS.

mharms commented:
We could rename as `_vn_RSS` as you suggested.  And should we just leave it sitting in `RssFolder`, _regardless_ of whether an `_hpXX_clean_vn` file exists in `ResultsFolder`?

glasserm commented:
Sure

mharms commented:
Ok, see c0a23ff.  I've renamed the file `_vn_RSS` within the matlab script itself, since @tbbrown needs to recompile it anyway.

glasserm commented:
Sounds good.

mharms commented:
@tbbrown I got the exact same error message as previously.  Did you update the compiler script with the needed change before recompiling?

```
Undefined function or variable 'save_avw'.
Error in RestingStateStats (line 143)
MATLAB:UndefinedFunction
```

tbbrown commented:
I made no changes myself. I merged the changes that were in the branch in github into my local branch and then did the compile. If the necessary changes were in the github version of the branch, then those changes should be in my local branch after the merge. I will investigate.

 Tim

On Tue, May 28, 2019, at 18:34, Michael Harms wrote:
> @tbbrown <https://github.com/tbbrown> I got the exact same error message as previously. Did you update the compiler script with the needed change before recompiling?

> `Undefined function or variable 'save_avw'.
Error in RestingStateStats (line 143)
MATLAB:UndefinedFunction
`
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Washington-University/HCPpipelines/pull/112?email_source=notifications&email_token=ABNXRR2JUNVSFKKBJFPDHF3PXW6QPA5CNFSM4HMSAVB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWNXRII#issuecomment-496728225>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABNXRRYHZEI6F2DGJWFA5HDPXW6QPANCNFSM4HMSAVBQ>.


--
Timothy B. Brown
Business & Technology Application Analyst III
Pipeline Developer (Human Connectome Project)
tbbrown(at)wustl.edu
________________________________________

The material in this message is private and may contain Protected Healthcare Information (PHI). 
If you are not the intended recipient, be advised that any unauthorized use, disclosure, copying 
or the taking of any action in reliance on the contents of this information is strictly prohibited. 
If you have received this email in error, please immediately notify the sender via telephone or 
return mail.

mharms commented:
I left that particular change to you to make, since I'm not that familar with the compiler stuff.  I suggested what I thought might be necessary at the end of this comment
https://github.com/Washington-University/HCPpipelines/pull/112#issuecomment-494153627

tbbrown commented:
I've added the `-a ${HCPPIPEDIR}/global/fsl/etc/matlab` line to the RestingStateStats compile script, did a recompile of just the RestingStateStats MATLAB code, and then checked-in and pushed the results to the `bugfix/compile_MSMALL_MATLAB_code` branch.

mharms commented:
This PR should now be ready to merge.  Following up on https://github.com/Washington-University/HCPpipelines/pull/112#issuecomment-494153627, in **compiled matlab mode** RSS now successfully runs to completion with `--bc-mode=NONE` or `--bc-mode=REVERT`, and `--wm=NONE` and `--csf=NONE`.  It also runs to completion, _and generates the expected pngs_, if the `--wm` and `--csf` flags are set to the `FreeSurfer{WM,CSF}RegLut.txt` files.  It _does_ take a long time to generate those pngs however (5.5 hours in one test run).  I don't know if interpreted matlab mode also takes that long for figure generation, or if that is instead indicative of some inefficiency in figure generation when working with compiled matlab.  Might be worth investigating that sometime, although I don't intend to pursue that at this particular time.
